src/app/clusters/ethernet_network_diagnostics_server/ethernet_network_diagnostics_server.cpp

```diff
@@ -16,11 +16,106 @@
  */
 
 #include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterceptor.h>
 #include <app/CommandHandler.h>
+#include <app/MessageDef/AttributeDataElement.h>
 #include <app/util/af.h>
+#include <app/util/attribute-storage.h>
+#include <lib/core/Optional.h>
+#include <platform/ConnectivityManager.h>
 
 using namespace chip;
+using namespace chip::app;
 using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::EthernetNetworkDiagnostics::Attributes;
+using chip::DeviceLayer::ConnectivityManager;
+
+namespace {
+
+class EthernetDiagosticsAttrInterceptor : public AttributeAccessInterceptor
+{
+public:
+    // Register for the EthernetNetworkDiagnostics cluster on all endpoints.
+    EthernetDiagosticsAttrInterceptor() :
+        AttributeAccessInterceptor(Optional<EndpointId>::Missing(), EthernetNetworkDiagnostics::Id)
+    {}
+
+    CHIP_ERROR ReadAttributeData(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter, bool * aDataRead) override;
+
+    bool IsRegistered() const { return sIsRegistered; }
+
+    void RegisterIfNeeded()
+    {
+        if (!sIsRegistered)
+        {
+            registerAttributeAccessInterceptor(this);
+        }
+        sIsRegistered = true;
+    }
+
+private:
+    static bool sIsRegistered;
+
+    CHIP_ERROR ReadDataIfSupported(CHIP_ERROR (ConnectivityManager::*getter)(uint64_t &), TLV::TLVWriter * aWriter);
+};
+
+EthernetDiagosticsAttrInterceptor gAttrInterceptor;
+bool EthernetDiagosticsAttrInterceptor::sIsRegistered = false;
+
+CHIP_ERROR EthernetDiagosticsAttrInterceptor::ReadAttributeData(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter,
+                                                                bool * aDataRead)
+{
+    if (aClusterInfo.mClusterId != EthernetNetworkDiagnostics::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    *aDataRead = true;
+    switch (aClusterInfo.mFieldId)
+    {
+    case Ids::PacketRxCount: {
+        return ReadDataIfSupported(&ConnectivityManager::GetEthPacketRxCount, aWriter);
+    }
+    case Ids::PacketTxCount: {
+        return ReadDataIfSupported(&ConnectivityManager::GetEthPacketTxCount, aWriter);
+    }
+    case Ids::TxErrCount: {
+        return ReadDataIfSupported(&ConnectivityManager::GetEthTxErrCount, aWriter);
+    }
+    case Ids::CollisionCount: {
+        return ReadDataIfSupported(&ConnectivityManager::GetEthCollisionCount, aWriter);
+    }
+    case Ids::OverrunCount: {
+        return ReadDataIfSupported(&ConnectivityManager::GetEthOverrunCount, aWriter);
+    }
+    default: {
+        *aDataRead = false;
+        break;
+    }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EthernetDiagosticsAttrInterceptor::ReadDataIfSupported(CHIP_ERROR (ConnectivityManager::*getter)(uint64_t &),
```

Comment:
It seems we assume we only read out uint64_t value from platform, that is true only for Ethernet mandatory attributes, what is the result is not uint64_t?

Reply:
Then the relevant getter on the connectivity manager will not match this signature, the attempt to use this helper for such an attribute will not compile, and either we will need to add a different helper for the relevant type or just hand-write the code as needed.

Reply:
I think this function is just a bit of convenience to reduce the boiler-plate above. For non uint64_t attributes, presumably, direct calls to the `Connectivitymgr()` will be made in the switch statement above.